nabs/plans.py

```diff
@@ -653,3 +659,138 @@ def daq_a3scan(m1, a1, b1, m2, a2, b2, m3, a3, b3, nsteps):
+
+
+def fixed_target_scan(sample, detectors, x_motor, y_motor, scan_motor, ss,
+                      n_shots, path, snake_like=True):
+    """
+    Scan over two variables in steps simultaneously.
+
+    This is a `nabs` version of ``bluesky``'s built-in
+    `bluesky.plans.list_scan` plan.
+    This scan is designed to be used with 3 motors, the x and y motors that
+    move to a designaget target on a sample wafer, and one scan_motor that can
+    be delay time, laser power or some other motor positions.
+
+    Parameters
+    ----------
+    sample : str
+        The name of the sample we're interested in.
+    detectors : list
+        Objects to read into Python in the scan.
+    x_motor : obj
+        Motor object corresponding to the x axes.
+    y_motor : obj
+        Motor object corresponding to the y axes.
+    scan_motor : obj
+        The motor being scanned. It can be e.g., delay time, laser power, some
+        other motor position, etc.
+    ss : list
+        List of all the points (samples) for the scan_motor to go through.
+    n_shots : int
+        Indicates how many shots should be taken, or how many samples should
+        be scanned on the grid.
+    path : str
+        Path where the sample file is located.
+    snake_like : bool, optional
+        Indicates if the scanning should be following a snake-like pattern.
+        Defaluts to `True`.
+    """
+
+    # TODO: remember what targets have been shot -
+    # display this in a little window - bit mapping?
+
+    global _last_index
+    detectors = list(detectors) + [scan_motor]
+
+    rows, columns, last_shot_index, xx, yy = get_sample_info(sample, path)
+    if snake_like:
+        xx = snake_grid_list(np.array(xx).reshape(rows, columns))
+
+    _last_index = last_shot_index
+    if (last_shot_index + (n_shots * len(ss))) >= len(xx):
+        raise IndexError('The number of last_shot_index + n_shots * len(ss): '
+                         f'{last_shot_index} + {n_shots} * {len(ss)}= '
+                         f'{last_shot_index + (n_shots * len(ss))}: '
+                         f' bigger than the available samples: {len(xx)}. '
+                         'Please provide a number in range.')
+
+    def inner_scan():
+        global _last_index
+        try:
+            yield from bps.open_run(md={})
+            for i in range(len(ss)):
+                yield from bps.mv(scan_motor, ss[i])
+                yield from bps.read(scan_motor)
+
+                x_pos = xx[(_last_index + 1):(
+                    (_last_index + 1) + n_shots)]
+                y_pos = yy[(_last_index + 1):(
+                    (_last_index + 1) + n_shots)]
+                yield from bpp.stub_wrapper(bp.list_scan(detectors, x_motor,
+                                            x_pos, y_motor, y_pos))
+
+                _last_index = _last_index + n_shots
+            yield from bps.close_run()
+            update_sample(sample, path, _last_index)
+        except Exception:
```

Comment:
`bluesky` does have some built-in tools for plans that do "something" after being aborted, see https://github.com/bluesky/bluesky/blob/4d8240e160ab0ba78652848c595e3b431a0c6175/bluesky/preprocessors.py#L469` and as an example https://github.com/bluesky/bluesky/blob/4d8240e160ab0ba78652848c595e3b431a0c6175/bluesky/preprocessors.py#L916

E.g. you definitely want to do `bps.close_run()` regardless of when the scan breaks, usually this would be done for you in https://github.com/bluesky/bluesky/blob/4d8240e160ab0ba78652848c595e3b431a0c6175/bluesky/preprocessors.py#L307

Reply:
Ohh I have looked at this `finalize_wrapper` a bunch of times before and did not think that it does what it does - that's very nice - thank you for sharing these.

I am using now in 7ac3131 the `finalize_wrapper` and added a callable: `finalize()` to do the updating of the index. 
I am also using the `run_decorator` now, so I took out the extra `open_run`/`close_run` Msgs
.. hopefully I got your suggestions right - please let me know

Reply:
I think this looks very clean now

Reply:
Ohh one thing here that I noticed is that the `summarize_plan` will change the `last_shot_index` as well.... do scientists use this functionality usually?

Reply:
I doubt this is used as much as it should be, but it's worth considering how to deal with it

Reply:
ohh i agree - I just can't think of how I would do that... 

Please let me know if you have anything in mind.

From my scan I can't know if `summarize_plan` will be called - at least I don't know how I would do that. 

If I could magically transform my `update_sample` method to be like a `Msg.command` - but even then, I don't know how I would tell my function not to do anything when `summarize_plan` is called....

I will do more research here...

Reply:
I can do something like this:
```
   import ast
    import inspect
    call_names = [c.func.id for c in ast.walk(ast.parse(inspect.getsource(summarize_plan)))
              if isinstance(c, ast.Call)]
    if 'print_summary_wrapper' in call_names:
        print('FOUND IT..........')
```
and go from there... but I have not used this before and I don't know how save that would be to include in a scan function...

Reply:
Yeah I'm also not sure the "best" way to accomplish this. To make the `bluesky` scan work properly in `summarize_plan`, ALL operations that make some change must be done as messages... Your approach seems like it would work to me, but I'm not really sure if we should be inspecting the call stack as part of a scan.

An alternative might be something like, always using `Signal` objects to hold values and write out files, etc... I'm not sure what built-in tooling exists for this. Any thoughts @klauer?

At any rate, this doesn't need to be a priority if it's holding up the project.

Reply:
Looks like I missed the call-out on this one - sorry!

I think the proper approach is as Zach mentioned: such a thing likely should be a custom `Msg`. The summary would only then list the message and arguments.

That said, stashing away information in a Python-only signal is pretty useful, too. It exposes the state in a way that can easily be introspected.

Reply:
Ok sounds good. I think I'll go the custom `Msg` route (just because I might be able to find more info on how to do this...)
For this pass I'll just have to make sure I tell the scientists not to use the `summarize_plan` nor the `plot_raster_path`.....

Reply:
Custom messages have their own quirks to be aware of- mainly, you need to teach the RunEngine how to interpret them, adding some startup configuration cost.